app/controllers/admin/badges_controller.rb

```diff
@@ -52,37 +54,79 @@ def mass_award
     end
 
     replace_badge_owners = params[:replace_badge_owners] == 'true'
-    BadgeGranter.revoke_all(badge) if replace_badge_owners
+    grant_existing_holders = params[:grant_existing_holders] == 'true'
+    if grant_existing_holders && !badge.multiple_grant?
+      render_json_error(
+        I18n.t('badges.mass_award.errors.cant_grant_multiple_times', badge_name: badge.display_name),
+        status: 422
+      )
+      return
+    end
 
-    batch_number = 1
     line_number = 1
-    batch = []
-
+    entries = []
     File.open(csv_file) do |csv|
-      mode = Email.is_valid?(CSV.parse_line(csv.first).first) ? 'email' : 'username'
-      csv.rewind
+      csv.each_line do |line|
+        line = CSV.parse_line(line).first&.strip&.downcase
+        line_number += 1
 
-      csv.each_line do |email_line|
-        line = CSV.parse_line(email_line).first
+        entries << line if line.present?
 
-        if line.present?
-          batch << line
-          line_number += 1
+        if entries.size > MAX_CSV_LINES
+          return render_json_error I18n.t('badges.mass_award.errors.too_many_csv_entries', count: MAX_CSV_LINES), status: 400
         end
+      end
+    end
 
-        # Split the emails in batches of 200 elements.
-        full_batch = csv.lineno % (BadgeGranter::MAX_ITEMS_FOR_DELTA * batch_number) == 0
-        last_batch_item = full_batch || csv.eof?
+    BadgeGranter.revoke_all(badge) if replace_badge_owners
 
-        if last_batch_item
-          Jobs.enqueue(:mass_award_badge, users_batch: batch, badge_id: badge.id, mode: mode)
-          batch = []
-          batch_number += 1
-        end
+    usernames = []
+    emails = []
+    entries.each do |entry|
+      if entry.include?('@')
+        emails << entry
+      else
+        usernames << entry
       end
     end
 
-    head :ok
+    usernames_map_to_ids = emails_map_to_ids = {}
```

Comment:
This doesn't feel right to me. Both variables are assigned to the same object. If the CSV has a mix of usernames and emails, we'll actually discard all the usernames.

Reply:
Right, they do share the same object here, but later in the lines directly after this one the variables are reassigned to different objects so if there are emails or usernames, they'll end up referencing different objects. I did this to make the code 1 line shorter, but I guess it's not really worth it 😅 Will split this into 2 lines.